### PR TITLE
MCOL-3769 This commit divides create_SH in two parts: before rewrites…

### PR DESCRIPTION
**Pull request by drrtuy**
This commit divides create_SH in two parts : before rewrites and 
after rewrites. We can safely fallback from SH to DH and table mode from
'before rewrites' part. We also returns meaningful warning.

### CODE REVIEW

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 1]
```diff
@@ -759,65 +759,36 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
     // MCOL-2178 Disable SP support in the select_handler for now.
     // Check the session variable value to enable/disable use of
     // select_handler
+    if (!get_select_handler(thd))
+    {
+        return handler;
+    }
+
+    // Disable SP support in the select_handler for now.
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp:766]: This comment is duplicated above.

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 2]
```diff
@@ -830,31 +801,55 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
...
         }
     }
 
-    if (!unsupported_feature)
+    // We apply dedicated rewrites from MDB here so MDB's data structures
+    // becomes dirty and CS has to raise an error in case of any problem
+    // or unsupported feature.
+    handler= new ha_columnstore_select_handler(thd, select_lex);
+    JOIN *join= select_lex->join;
     {
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp:813]: Do we need these braces now that the if condition is gone?

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 3]
```diff
@@ -830,31 +801,55 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
...
-        int rc= 0;
-        // Skip execution for EXPLAIN queries
-        if (!thd->lex->describe)
+        disable_indices_for_CEJ(thd);
+
+        if (select_lex->handle_derived(thd->lex, DT_MERGE))
         {
-            rc= ha_cs_impl_pushdown_init(&mhi, handler->table);
+            unsupported_feature = true;
+            handler->err_msg.assign("create_columnstore_select_handler(): Internal error happens in SL::handle_derived()");
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp]: Instead of happens, occurred would be a better word here.

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 4]
```diff
@@ -830,31 +801,55 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
...
+                join->conds, TRUE, FALSE);
+        }
+        // MCOL-3747 IN-TO-EXISTS rewrite inside MDB didn't add
+        // an equi-JOIN condition.
+        if (!unsupported_feature
+            && optimize_unflattened_subqueries_mcs(select_lex))
+        {
             unsupported_feature = true;
-        return handler;
+            handler->err_msg.assign("create_columnstore_select_handler: Internal error happens in optimize_unflattened_subqueries_mcs()");
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp]: Instead of happens, occurred would be a better word here.

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 5]
```diff
@@ -868,7 +863,8 @@ ha_columnstore_select_handler::ha_columnstore_select_handler(THD *thd,
                                                              SELECT_LEX* select_lex)
   : select_handler(thd, mcs_hton)
 {
-  select = select_lex;
+  select= select_lex;
+  rewrite_error= 0;
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp]: should be = false

File: dbcon/mysql/ha_mcs_pushdown.cpp [hunk 6]
```diff
@@ -892,10 +888,26 @@ int ha_columnstore_select_handler::init_scan()
...
 
-    // Dummy init for SH. Actual init happens in create_SH
-    // to allow fallback to other handlers if SH fails.
     int rc = 0;
 
+    if (!rewrite_error)
+    {
+        // handler::table is the place for the result set
+        // Skip execution for EXPLAIN queries
+        if (!thd->lex->describe)
```
Comment (tntnatbry) [inline dbcon/mysql/ha_mcs_pushdown.cpp:898]: Curious, why was this moved from create_SH to here?

**Review by tntnatbry — CHANGES_REQUESTED** [review]

File: dbcon/mysql/ha_mcs_pushdown.cpp (hunk 2 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_pushdown.cpp:813]: We do. This is context where we guard variables we don't need outside.

File: dbcon/mysql/ha_mcs_pushdown.cpp (hunk 6 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_pushdown.cpp:898]: Good question.
B/c create_SH is inappropriate place to execute the query(we can't even return and error w/o breaking the execution and init_SH is the place to execute the query.
I placed execution in create_SH in the first place to have a fallback but later understood it could break execution when we partially rewrote SELECT_LEX and then fallback to table API

File: dbcon/mysql/ha_mcs_pushdown.cpp (hunk 1 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_pushdown.cpp:766]: Thanks to rebase that might be error prone.

File: dbcon/mysql/ha_mcs_pushdown.cpp (hunk 4 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_pushdown.cpp]: To be honest it at this point the error happens for real not occured :)

**Review by pleblanc1976 — CHANGES_REQUESTED** [review]
It looks good to me, but I don't know this code very well.  If it doesn't break anything in our regression test I'm OK with it.  Gagan can take the lead on this one.

**Comment by drrtuy** [comment]
All relevant comments had been fixed.

**Review by tntnatbry — APPROVED** [review]

**Comment by tntnatbry** [comment]
There is compilation errors in BB. Can you please rebase on top of current develop?

**Comment by drrtuy** [comment]
Rebased.

**Review by tntnatbry — APPROVED** [review]

**Review by pleblanc1976 — APPROVED** [review]
